pkg/fastesturl/fastesturl.go

```diff
@@ -0,0 +1,107 @@
+package fastesturl
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// ReqCheckFunc checks if a *http.Response
+// is valid for returning to a client
+type RespCheck func(*http.Response) bool
+
+// FatestURL implements a method for obtaining the first
+// *http.Response object returned from a list of URLs.
+type FastestURL struct {
+	// an http client to use for all requests
+	Client *http.Client
+	// a list of urls we will concurrecntly attempt to request
+	URLs []*url.URL
+	// a template request object we will copy fields from
+	Request *http.Request
+	// a function provided by the caller to determine if a request
+	// should be returned
+	RespCheck RespCheck
+}
+
+// New is a constructor for a FastestURL
+func New(client *http.Client, req *http.Request, check RespCheck, urls []*url.URL) *FastestURL {
+	if client == nil {
+		client = &http.Client{}
+	}
+	if req == nil {
+		req = &http.Request{}
+	}
+	if check == nil {
+		check = func(resp *http.Response) bool {
+			if resp.StatusCode == 200 {
+				return true
+			}
+			return false
+		}
+	}
+	return &FastestURL{
+		Client:    client,
+		URLs:      urls,
+		Request:   req,
+		RespCheck: check,
+	}
+}
+
+// Do will return the first *http.Response which passes
+// f.RespCheck.
+//
+// If no successful *http.Response is obtained a nil is returned
+func (f *FastestURL) Do(ctx context.Context) *http.Response {
+	log := zerolog.Ctx(ctx).With().Str("routine", "fasesturl_do").Logger()
+	cond := sync.NewCond(&sync.Mutex{})
+	var response *http.Response
+	tctx, cancel := context.WithTimeout(ctx, 30*time.Second)
```

Comment:
@hdonnay 

What do you think is the right way to handle this timeout? Should it remain constant

Reply:
I think the caller should handle it with the passed-in context

Reply:
cool, I won't derive a ctx at all then and assume the one being passed in will be used as the caller defines.